### ISSUE TITLE
YOLOE Visual Prompt ROI Discrepancy

### ISSUE THREAD

**Issue by 102757017** [ultralytics/ultralytics#22233]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

When re-detecting on the reference image itself. Why Detected Bounding Boxes Differ from Prompted ROI and the confidence is very low?

### Additional

When using YOLOE with visual prompts, I observed that the detected object's bounding box significantly differs from and encompasses a larger area than the ROI specified in visual_prompts, even when re-detecting on the reference image itself. This behavior is  noticeable for a circular object where the manually defined ROI intended to tightly enclose it.
For instance, I provided the following visual_prompts:
`
import numpy as np
import torch
import torch.nn.functional as F  
from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOEVPSegPredictor,YOLOEVPDetectPredictor

model = YOLOE("yoloe-v8s-seg.pt")
visual_prompts = dict(
    bboxes=np.array(
        [
            [37.21,40.64,101.2,102.7],  
        ],
    ),
    cls=np.array(
        [
            0,  
        ]
    ),
)

results = model.predict(
    "boss.jpg",  
    refer_image="boss.jpg",  
    visual_prompts=visual_prompts,
    predictor=YOLOEVPSegPredictor,
    #save=True,
    exist_ok=True,
)
results[0].show(labels=False)
`
the resulting bounding box for the detected object often includes more surrounding area than the ROI, as illustrated in the image below. and the confidence is very low

![Image](https://github.com/user-attachments/assets/d7b5fc87-8d4a-47ff-96f0-8fa72e3e47d1)

![Image](https://github.com/user-attachments/assets/c680ba3f-5698-47a7-92a8-7f3257f215e7)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @102757017, thank you for your detailed report and for using Ultralytics 🚀! This is an automated response to help you get support faster. An Ultralytics engineer will follow up here soon.

Your description about YOLOE visual prompts producing larger-than-ROI detections with low confidence is noted. To help us investigate efficiently:

- If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) that anyone can run end-to-end
- Include: exact `ultralytics` package version, PyTorch version, Python version, OS, device (CPU/GPU + CUDA/CUDNN), the exact model weights path you used, the precise `model.predict(...)` arguments, and a single small test image (or synthetic image) with your `visual_prompts` that reproduces the issue
- Share the full console output/logs and a screenshot of the resulting visualization for the MRE

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).  

For general usage, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples and where many common questions may already be answered.

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/)
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community

## Upgrade

Before we dig in, please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify the behavior persists on the newest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the thorough context and code snippet. Please share the requested MRE details so we can reproduce the ROI discrepancy and low confidence behavior on our side and assist you promptly 🙏

**Comment by Y-T-G** [comment]
You can use the larger models for better results. Or add more prompts.

**Comment by Y-T-G** [comment]
Also your object prompts are very small. It's not going to be precise due to downsampling.

**Comment by 102757017** [comment]
> Also your object prompts are very small. It's not going to be precise due to downsampling.

"I tried using yoloe-11l-seg.pt, and the results were still the same.

Also, I did not specify "imgsz" in my code, so its default value is 640. When extracting visual_prompts, the reference image should be upscaled to 640 before extracting the VPE. For prediction, the reference image should also be upscaled to 640 before making the prediction. This process should involve upsampling, not downsampling, correct?"

**Comment by glenn-jocher** [comment]
Great question—`imgsz` defaults to 640, but YOLOE uses letterbox resize+pad, so images larger than 640 are downsampled and smaller are upsampled; then `pre_transform` maps your prompt boxes into the letterboxed space and builds low‑resolution visual maps (~1/8 scale), which is why very small/tight ROIs can look expanded and score low. See the YOLOE predictor pre_transform docs for details and the LoadVisualPrompt reference on the 1/8 scaling behavior: [YOLOE predictor reference (pre_transform)](https://docs.ultralytics.com/reference/models/yolo/yoloe/predict/) and [LoadVisualPrompt](https://docs.ultralytics.com/reference/data/augment/).

Two quick tweaks that usually help small objects:
- Increase input size and give a slightly larger prompt box (a small margin around the object) or a couple of examples.
  
Example:
```python
from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOEVPSegPredictor
import numpy as np

model = YOLOE("yoloe-11l-seg.pt")
visual_prompts = dict(
    bboxes=np.array([[37.21, 40.64, 101.2, 102.7]]),  # consider a bit larger box
    cls=np.array([0]),
)
results = model.predict(
    "boss.jpg",
    refer_image="boss.jpg",
    visual_prompts=visual_prompts,
    predictor=YOLOEVPSegPredictor,
    imgsz=1280,  # higher resolution to preserve small ROI detail
)
```

If this still looks off on the same `boss.jpg`, please attach that image plus the exact `visual_prompts` array so we can reproduce on our side. For more usage notes, the visual/text prompt workflow is summarized in [YOLOE predict usage](https://docs.ultralytics.com/models/yoloe/#predict-usage).

**Comment by 102757017** [comment]
Inconsistent imgsz behavior between VPE extraction and object detection inference stages in YOLOE.

Observations and Findings:

**1. Anomalous Behavior with imgsz=1280:**

- When imgsz was set to 1280 and a slightly larger bounding box prompt (with a small margin around the object) was provided, the model failed to detect any objects. This was contrary to expectations.

**2. Potential Impact of imgsz on VPE Extraction:**

- Preliminary validation suggests that the imgsz parameter within the model.predict() method appears to only affect the image size fed into the model during the inference stage.
- For Visual Prompt Embedding (VPE) extraction, even when imgsz is specified in model.predict(), the VPE extraction process (e.g., when model.predictor.get_vpe() is called) seems to still use a default image size of 640 for processing the reference image.

**3.Issues Arising from Mismatched imgsz between VPE Extraction and Inference:**

- Scenario A (Manual Feature Extractor and Inference imgsz Match):
    - When manually using the feature extractor (model.predictor) and specifying imgsz=576 for VPE extraction.
    - And concurrently specifying imgsz=576 for object detection inference.
    - In this case, the detected regions were normal and accurate.
- Scenario B (Overall model.predict Call):
    - When performing detection using results = model.predict("boss.jpg", refer_image="boss.jpg", visual_prompts=visual_prompts, predictor=YOLOEVPSegPredictor, imgsz=576).
    - The detected object regions were incorrect, showing a discrepancy compared to Scenario A.

**Key Questions:**

- Is the issue in Scenario B precisely because the imgsz parameter in model.predict (while affecting the final inference) is not being propagated to the VPE extraction stage, causing the VPE to still be extracted based on a 640-sized image?
- If this is indeed the case, it implies:
     - For reference images larger than 640: Even if imgsz is set larger, VPE extraction might still downsample it to 640 before VPE is extracted. This could lead to VPE losing some high-resolution information, although the inference image will then be upscaled to the specified imgsz.
    - For reference images smaller than 640: VPE extraction would upscale it to 640. However, if the inference stage's imgsz is also smaller than 640 (e.g., 576), then the VPE is extracted from an image upscaled to 640, while inference is performed on an image upscaled to 576. This would result in a mismatch between the VPE scale and the inference image scale, severely impacting detection performance.

_ = model.predict(
    ref_image1,
    visual_prompts=visual_prompts_image1,
    predictor=YOLOEVPDetectPredictor,
    imgsz=576,   #specifying imgsz for VPE extraction
    save=False,
    verbose=False
)
visual_prompts_image1 = dict(
    bboxes=np.array([[37.21,40.64,101.2,102.7]]), 
    cls=np.array([1])
)
model.predictor.set_prompts(visual_prompts_image1)
vpe_class0_final = model.predictor.get_vpe("boss.jpg" )

**Comment by Y-T-G** [comment]
@102757017 Can you try 
```
pip uninstall ultralytics -y && pip install git+https://github.com/ultralytics/ultralytics@main
```

**Comment by UltralyticsAssistant** [comment]
Good news — a potential fix for this issue has been merged in PR [YOLOE: Use passed imgsz for visual prompt extraction](https://github.com/ultralytics/ultralytics/pull/22238).

What changed and why this may resolve the issue:
- Predict now propagates the provided `imgsz` into export args, so exported models (ONNX/TensorRT, etc.) use the exact image size you requested rather than a default. This helps prevent subtle mismatches that can impact accuracy, speed, and memory.
- YOLO26 docs were refreshed for clarity. See the updated guide: [YOLO26 model overview and previews](https://docs.ultralytics.com/models/yolo26/).

Thanks to the author Y-T-G and contributor Laughing-q for the work here!

How to test the fix:
- Install the latest from main:
  - `pip install -U pip`
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Run a quick check to confirm `imgsz` is respected during predict-triggered export:
  ```python
  from ultralytics import YOLO

  model = YOLO('yoloe.pt')  # example model
  results = model.predict(source='path/to/image.jpg', imgsz=640, save=False, export=True, format='onnx')
  print('Export complete; verify the ONNX input shape matches imgsz=640.')
  ```
- Alternatively, wait for the next official PyPI release and upgrade then:
  - `pip install -U ultralytics`

Feedback welcome on whether this resolves the issue in your environment. Thanks for reporting this, and please share details if anything persists so it can be refined further. 🙏
